project_task_autocreate/models/project_task_template.py

```diff
@@ -0,0 +1,437 @@
+
+    @api.constrains(
+        'model_id', 'use_relative_deadline',
+        'relative_deadline', 'relative_deadline_op',
+        'relative_deadline_units', 'relative_deadline_delta',
+    )
+    def _check_relative_deadline(self):
+        for rec in self:
+            if not rec.use_relative_deadline:
+                continue
+            if not all([rec.relative_deadline_op,
+                        rec.relative_deadline_units,
+                        rec.relative_deadline_delta]):
+                raise ValidationError(_(
+                    "All fields of the relative deadline must be set."
+                ))
+            relstr = rec.relative_deadline
+            if relstr != 'today':
+                rec._check_relative_field(relstr, 'Deadline', 'date')
+
+    @api.constrains('model_id', 'rule_ids')
+    def _check_model_id(self):
+        for rec in self:
+            if rec.rule_ids and not rec.model_id:
+                raise ValidationError(_(
+                    "Templates with rules must have a model set."
+                ))
+
+    @api.multi
+    def write(self, vals):
+        # Update records before creating tasks
+        res = super(ProjectTaskTemplate, self).write(vals)
+        if 'trigger_record_id' in vals:
+            rec_id = vals['trigger_record_id']
+            models = set(self.mapped('model'))
+            if len(models) > 1:
+                raise UserError(_(
+                    "Task creation cannot be triggered for multiple templates "
+                    "with different models at the same time."
+                ))
+            rec = self.env[models.pop()].browse(rec_id)
+            self.create_tasks_from_records(rec)
+        return res
+
+    @api.multi
+    @api.returns('project.task')
+    def create_task(self):
+        """
+        Create a task from the template.
+        """
+        tasks = self.env['project.task']
+        for template in self:
+            tasks |= tasks.create(
+                template.get_task_vals()
+            )
+        return tasks
+
+    @api.multi
+    @api.returns('project.task')
+    def create_tasks_from_records(self, records):
+        """
+        Create a task from the template for each record in records.
+        """
+        tasks = self.env['project.task']
+        for template in self:
+            if records._name != template.model:
+                raise UserError(_(
+                    "The template '%(template)s' can only be used with "
+                    "%(expected_model)s, not with %(model)s."
+                ) % {
+                    'template': template.name,
+                    'expected_model': template.model,
+                    'model': records._name,
+                })
+            for rec in records:
+                tasks |= tasks.create(
+                    template.get_task_vals(rec)
+                )
+        return tasks
+
+    @api.multi
+    def action_create_task(self):
```

Comment:
ça sert à quoi cette méthode. C'est quoi son utilité d'un point de vue fonctionnel?
David et Quentin n'ont pas d'idée aussi.

Reply:
@rimbendhaou who are you asking ?

Reply:
Joao. Est ce qu'il faut mettre @Joao?

Reply:
yes, specify who you are talking to. And because this is a public github repo, you should write in english.